### PR TITLE
Add ChangeUnitConstructor annotation to allow mongock to pick constructors for dependency injection

### PR DESCRIPTION
**Pull request by thekaleidoscope**
## Issue reference

[508](https://github.com/mongock/mongock/issues/508)

## Documentation PR reference

[38](https://github.com/mongock/mongock-docs/pull/38)

## Description and context

Used Optionals and Stream for null safe iterations

## Benefits

[What benefits will be realized by the code change?]

Improve existing exceptions, add ability to specify which constructor to inject dependencies to.

## Possible Drawbacks

[What are the possible side-effects or negative impacts of the code change?]

None I can think of.

## Checklist 
- [x] Unit tests provided
- [ ] Integration tests provided 
- [x] Documentation updated in [documentation project](https://github.com/mongock/mongock-docs)
- [ ] Updated example in [example projects](https://github.com/mongock/mongock-examples) // can't find use of constructor there

### CODE REVIEW

**Comment by dieppa** [comment]
We'll take a look today ;).

Thanks

File: mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java
Comment (dieppa) [inline mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java]: It's fine but I don't see the need of using streams, plus we are creating it twice.
What about:
```java
  private Constructor<?> findDefaultConstructor(Class<?> type) {
    Constructor<?>[] constructors = type.getConstructors();
    if (constructors.length > 1) {
      logger.warn("Mongock found multiple constructors for changeUnit[{}]. " +
          "It's recommended to annotate the one you want Mongock to use with @ChangeUnitConstructor. " +
          "FROM VERSION 6 THIS WILL CAUSE AN ERROR ", type.getName());
    } else if(constructors.length == 0){
      throw new MongockException("Mongock cannot find a valid constructor for changeUnit[%s]", type.getName());
    }
    return constructors[0];
  }
```

File: mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java
Comment (dieppa) [inline mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java:153]: It's also fine, but there is no need for Supplier. 
What about:
```java
  private Optional<Constructor<?>> findChangeUnitConstructor(Class<?> type) {
    Stream<Constructor<?>> constructorStream = Arrays.stream(type.getConstructors())
        .filter(constructor -> constructor.isAnnotationPresent(ChangeUnitConstructor.class));
    if (constructorStream.count() > 1) {
      throw new MongockException("Found multiple constructors for changeUnit[%s] without annotation @ChangeUnitConstructor." +
          " Annotate the one you want Mongock to use to instantiate your changeUnit", type.getName());
    }
    return constructorStream.findFirst();
  }
```

**Review by dieppa — CHANGES_REQUESTED** [review]
Thanks! looks good, just a couple of minor changes before merging 😃

File: mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java
Comment (thekaleidoscope) [reply mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java]: The use of stream was done to keep a functional style of code for easier readability, It also supports Optionals out of box with functions like findFirst(), orElseThrow()

Since the number of constructors usually is a single digit number, I thought the tradeoff of readability trumped efficiency.

This code also followed the [Object Calisthenics](https://williamdurand.fr/2013/06/03/object-calisthenics/) rules like [Only One Level Of Indentation Per Method](https://williamdurand.fr/2013/06/03/object-calisthenics/#1-only-one-level-of-indentation-per-method) and 
[Don’t Use The ELSE Keyword](https://williamdurand.fr/2013/06/03/object-calisthenics/#2-dont-use-the-else-keyword)

Please let me know if the change is still recommended.

File: mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java
Comment (thekaleidoscope) [reply mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java:153]: This won't work as the `(constructorStream.count()`  ends stream. I have tests that'd fail with the recommended changes.

Comment (dieppa) [reply mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java:153]: True, my bad. leave it like it was

File: mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java
Comment (dieppa) [reply mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java]: I see your point, but still see more practical this approach.  Like it simple, from my point of view, the use of wrappers is good until they are not really needed. In this case I don\t see the need 😃

Comment (thekaleidoscope) [reply mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java]: okay, cool, will make the required changes

Comment (dieppa) [reply mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java]: Anyway, if you are not happy with else, which I don't see it wrong in a small method like this, you can do this, if you prefer:

```java
  private Constructor<?> findDefaultConstructor(Class<?> type) {
    Constructor<?>[] constructors = type.getConstructors();
    if(constructors.length == 0){
      throw new MongockException("Mongock cannot find a valid constructor for changeUnit[%s]", type.getName());
    }
    if (constructors.length > 1) {
      logger.warn("Mongock found multiple constructors for changeUnit[{}]. " +
          "It's recommended to annotate the one you want Mongock to use with @ChangeUnitConstructor. " +
          "FROM VERSION 6 THIS WILL CAUSE AN ERROR ", type.getName());
    }
    return constructors[0];
  }

```

**Up to you, I don't really mind**

**Review by dieppa — APPROVED** [review]

File: mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java
Comment (thekaleidoscope) [reply mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java]: done

**Review by osantana85 — APPROVED** [review]

**Comment by osantana85** [comment]
@thekaleidoscope we have approved this PR, so now you can merge it. Thanks for collaborating with Mongock!

**Comment by thekaleidoscope** [comment]
@osantana85 Thanks, I don't see how to do it though, It says `Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests.` In the past I saw @dieppa had merged the MRs

**Comment by osantana85** [comment]
@thekaleidoscope Ok, no problem, it's just about permissions. I'll merge it. Thanks.

**Pull request merged**
